lib/nolibc: Add README.md file (#778)
unikraft/unikraft: lib/nolibc/README.md

# nolibc, Unikraft's Lightweight Subset of the C Standard Library

> nolibc provides the basic core functionalities of libc in a lighter

Reviewer: ```suggestion
The `nolibc` library provides the basic core functionalities of libc in a lighter
```

> nolibc provides the basic core functionalities of libc in a lighter
> and more compact package, fulfilling the needs of most simple applications
> without occupying a large amount of memory, while also remaining POSIX-compliant.

Reviewer: ```suggestion
nolibc provides the basic core functionalities of libc in a lighter and more compact package, fulfilling the needs of most simple applications without occupying a large amount of memory, while also remaining POSIX-compliant.
```
Please use one sentence per line and one line per sentence, since it's easier to review.

> More complex applications that require a more comprehensive set of
> standard C
> functions are better suited with Unikraft's port of
> [musl](https://github.com/unikraft/lib-musl).

Reviewer: ```suggestion
More complex applications that require a more comprehensive set of standard C functions are better suited with Unikraft's port of [musl](https://github.com/unikraft/lib-musl).
```

> # Components

Reviewer: ```suggestion
## Components
```

> nolibc exposes the following symbols for external use:

Reviewer: ```suggestion
`nolibc` exposes the following symbols for external use:
```

> <details>

Reviewer: I would try to avoid using inline html as much as possible, you can just do a markdown list using `*` and add links to the man pages.
```md
* [function-name](link-to-man-page)
```

<li><tt>strncat</tt></li>
</ul>
</details>

<details>
<summary> strsignal </summary>
<br>
<a href="https://man7.org/linux/man-pages/man3/strsignal.3.html"> Official documentation</a>
<ul>
<li><tt>strsignal</tt></li>
</ul>
</details>

<details>
<summary> psignal </summary>
<br>
<a href="https://man7.org/linux/man-pages/man3/psignal.3.html"> Official documentation</a>
<ul>
<li><tt>psignal</tt></li>
</ul>
</details>

<details>
<summary> ctype </summary>
<br>
<a href="https://man7.org/linux/man-pages/man0/ctype.h.0p.html"> Official documentation</a>
<ul>
<li><tt>_nolibc_ctype</tt></li>
</ul>
</details>

<details>
<summary> errno </summary>
<br>
<a href="https://man7.org/linux/man-pages/man3/errno.3.html"> Official documentation</a>
<ul>
<li><tt>errno</tt></li>
</ul>
</details>

<details>
<summary> network </summary>
<br>
<a href="https://man7.org/linux/man-pages/man3/htons.3p.html"> Official documentation for <tt>htons</tt> and <tt>nthos</tt></a>
<br>
<a href="https://man7.org/linux/man-pages/man3/gethostbyname.3.html"> Official documentation for <tt>h_errno</tt></a>
<ul>
<li><tt>htons</tt></li>
<li><tt>nthos</tt></li>
<li><tt>h_errno</tt></li>
</ul>
</details>

> 
> 

Reviewer: ```suggestion

```

> # Limitations and caveats

Reviewer: ```suggestion
## Limitations and Caveats
```

> The small size and simple interface of nolibc came at the cost of
> reducing the scope of some functions. Notable limitations include:

Reviewer: ```suggestion
The small size and simple interface of nolibc came at the cost of reducing the scope of some functions.
Notable limitations include:
```

> * `fprintf()` being suitable only for `stdout` and `stderr`, not
> other files or devices.

Reviewer: ```suggestion
* `fprintf()` being suitable only for `stdout` and `stderr`, not other files or devices.
```

# nolibc, Unikraft's Lightweight Subset of the C Standard Library

The `nolibc` library provides the basic core functionalities of libc in a lighter and more compact package, fulfilling the needs of most simple applications without occupying a large amount of memory, while also remaining POSIX-compliant.

> More complex applications that require a more comprehensive set of standard C functions are better suited with Unikraft's port of [musl](https://github.com/unikraft/lib-musl).

Reviewer: ```suggestion
Complex applications that require a more comprehensive set of standard C functions are better suited with Unikraft's port of [musl](https://github.com/unikraft/lib-musl).
```

## Configuring Applications to Use `nolibc`

The `nolibc` library is an internal library of Unikraft and is enabled by default when configuring a new project.

> Opting for a different implementation of the C standard library will automatically override `nolibc` and utilise the one provided by the user.

Reviewer: ```suggestion
Opting for a different implementation of the C standard library will automatically override `nolibc` and utilize the one provided by the user.
```
Let's go for US English.

> This behaviour can be observed in the configuration menu under the `Library Configuration` section.

Reviewer: ```suggestion
This behavior can be observed in the configuration menu under the `Library Configuration` section.
```

> An example of a simple application that uses the `nolibc` library is the following:

Reviewer: ```suggestion
An example of a simple application that uses the `nolibc` library is the following:

```

```c
#include <stdio.h>
#include <string.h>

int main(int argc, char *argv[])
{
    char str[1024], *p;

    strcpy(str, "There's the way it ought to be");
    printf("%s\n", str);

    strcat(str, "and there's the way it is");
    p = strchr(str, 'b');
    p += 2;
    fprintf(stderr, "%s\n", p);

    return 0;
}
```

> Building and running the program above results in output similar to the following:

Reviewer: Please add some short steps on how to configure, build and run the app, similar to the [`ramfs` README file](https://github.com/unikraft/unikraft/tree/staging/lib/ramfs#configuring-applications-to-use-ramfs).

```console
SeaBIOS (version 1.13.0-1ubuntu1.1)
Booting from ROM...
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
             Epimetheus 0.12.0~4c7352c0
There's the way it ought to be
and there's the way it is
Console terminated, terminating guest (PID: 10490)...
```

## Limitations and Caveats

> The small size and simple interface of nolibc came at the cost of
> reducing the scope of some functions. Notable limitations include:

Reviewer: ```suggestion
The small size and simple interface of `nolibc` came at the cost of reducing the scope of some functions. Notable limitations include:
```
Use one line per sentence and one sentence per line.

> *   `fprintf()` being suitable only for `stdout` and `stderr`, not other files or devices.

Reviewer: ```suggestion

*   `fprintf()` being suitable only for `stdout` and `stderr`, not other files or devices.
```

> *   `fprintf()` being suitable only for `stdout` and `stderr`, not other files or devices.
> *   no internal buffering for `stdio` functions.
> *   no implementation of `fscanf` and its derivatives (input should be received only from `stdin`).

Reviewer: ```suggestion
* `fprintf()` being suitable only for `stdout` and `stderr`, not other files or devices.
* no internal buffering for `stdio` functions.
* no implementation of `fscanf` and its derivatives (input should be received only from `stdin`).
```

# nolibc, Unikraft's Lightweight Subset of the C Standard Library

The `nolibc` library provides the basic core functionalities of libc in a lighter and more compact package, fulfilling the needs of most simple applications without occupying a large amount of memory, while also remaining POSIX-compliant.

> Complex applications that require a more comprehensive set of standard C functions are better suited with Unikraft's port of [musl](https://github.com/unikraft/lib-musl).

Reviewer: ```suggestion
Complex applications that require a more comprehensive set of standard C functions are better suited with Unikraft's port of [Musl](https://github.com/unikraft/lib-musl).
```

## Configuring Applications to Use `nolibc`

The `nolibc` library is an internal library of Unikraft and is enabled by default when configuring a new project.
Opting for a different implementation of the C standard library will automatically override `nolibc` and utilize the one provided by the user.
This behavior can be observed in the configuration menu under the `Library Configuration` section.

An example of a simple application that uses the `nolibc` library is the following:

```c
#include <stdio.h>
#include <string.h>

int main(int argc, char *argv[])
{
    char str[1024], *p;

    strcpy(str, "There's the way it ought to be");
    printf("%s\n", str);

    strcat(str, "and there's the way it is");
    p = strchr(str, 'b');
    p += 2;
    fprintf(stderr, "%s\n", p);

    return 0;
}
```

The steps to configure the application to use `nolibc` are as follows:

1. Enter the configuration interface with:

    ```console
    $ make menuconfig
    ```

1. Enter the `Library Configuration` submenu
1. Select `nolibc:` `Only necessary subset of libc functionality`

> 1. Choose between using Unikraft-specific assertions or libc assertions by toggling `Implement assertions with libukdebug`

Reviewer: Rephrase this so it does not sound like something the user should do, but as something that the user could do (i.e. something like `You can switch between [Unikraft-specific assertions](link to ukdebug assertion header file) or [nolibc assertions](link to the nolibc assert.h header file) by toggling the Implement assertions with libukdebug option`).

Building and running the program above results in output similar to the following:

```console
SeaBIOS (version 1.13.0-1ubuntu1.1)
Booting from ROM...
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
             Epimetheus 0.12.0~4c7352c0
There's the way it ought to be
and there's the way it is
Console terminated, terminating guest (PID: 10490)...
```

## Limitations and Caveats

> The small size and simple interface of nolibc came at the cost of reducing the scope of some functions. Notable limitations include:

Reviewer: ```suggestion
The small size and simple interface of `nolibc` came at the cost of reducing the scope of some functions.
Notable limitations include:
```

Complex applications that require a more comprehensive set of standard C functions are better suited with Unikraft's port of [Musl](https://github.com/unikraft/lib-musl).

## Configuring Applications to Use `nolibc`

The `nolibc` library is an internal library of Unikraft and is enabled by default when configuring a new project.
Opting for a different implementation of the C standard library will automatically override `nolibc` and utilize the one provided by the user.
This behavior can be observed in the configuration menu under the `Library Configuration` section.

An example of a simple application that uses the `nolibc` library is the following:

```c
#include <stdio.h>
#include <string.h>

int main(int argc, char *argv[])
{
    char str[1024], *p;

    strcpy(str, "There's the way it ought to be");
    printf("%s\n", str);

    strcat(str, "and there's the way it is");
    p = strchr(str, 'b');
    p += 2;
    fprintf(stderr, "%s\n", p);

    return 0;
}
```

> The steps to configure the application to use `nolibc` are as follows:

Reviewer: ```suggestion
The steps for configuring the application to use `nolibc` are as follows:
```

1. Enter the configuration interface with:

> 
>       ```console
>       $ make menuconfig

Reviewer: ```suggestion

```console
$ make menuconfig
```
No  need for extra indentation.

```

1. Enter the `Library Configuration` submenu

> 1. Select `nolibc:` `Only necessary subset of libc functionality`

Reviewer: ```suggestion
1. Select `nolibc: Only necessary subset of libc functionality`
```

> 1. This submenu allows you to choose between using [Unikraft-specific assertions](../ukdebug/include/uk/assert.h) or [`nolibc` assertions](./include/assert.h) by toggling the `Implement assertions with libukdebug` option

Reviewer: Add full links to the files (https://github.com/unikraft...), so they don't get messed up when rendering them on the docs website.

Building and running the program above results in output similar to the following:

```console
SeaBIOS (version 1.13.0-1ubuntu1.1)
Booting from ROM...
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
             Epimetheus 0.12.0~4c7352c0
There's the way it ought to be
and there's the way it is
Console terminated, terminating guest (PID: 10490)...
```

## Limitations and Caveats

> The small size and simple interface of nolibc came at the cost of reducing the scope of some functions.

Reviewer: ```suggestion
The small size and simple interface of `nolibc` came at the cost of reducing the scope of some functions.
```

Notable limitations include:

> - `fprintf()` being suitable only for `stdout` and `stderr`, not other files or devices.

Reviewer: ```suggestion
- `fprintf()` being suitable only for `stdout` and `stderr`, not other files or devices
```
Only add dots at the end of full sentences.

> - no internal buffering for `stdio` functions.

Reviewer: ```suggestion
- No internal buffering for `stdio` functions
```

> - no implementation of `fscanf` and its derivatives (input should be received only from `stdin`).

Reviewer: ```suggestion
- No implementation of `fscanf` and its derivatives (input should be received only from `stdin`)
```
